### PR TITLE
NF prints summary of its activity on shutdown

### PR DESCRIPTION
**Pull request by dennisafa**
We now print NF specific stats on shutdown, such as total RX, dropped, etc.

## Summary:
Adds the statistics print call before we exit the NF, giving a summary of its activity. An example of speed_tester + simple_forward is shown:
Screen Shot 2020-01-02 at 4 13 08 PM

**Usage:**

| This PR includes         |          |
| ------------------------ | -------- |
| Resolves issues          |  
| Breaking API changes     |
| Internal API changes     | X
| Usability improvements   |  
| Bug fixes                |
| New functionality        | X
| New NF/onvm_mgr args     | 
| Changes to starting NFs  |  
| Dependency updates       | 
| Web stats updates        | 

## Merging notes:
 - Dependencies: None  

**TODO before merging :**
 - [x] PR is ready for review

## Test Plan:
Chained various NF's and exited to verify the final output was correct.

## Review: 
@koolzz @kevindweb let me know if we should add/remove any stats.

### CODE REVIEW

**Comment by onvm** [comment]
> In response to PR creation

## CI Message

Your results will arrive shortly

**Comment by dennisafa** [comment]
> Code looks clean and testing worked as usual! Good job working on this @dennisafa, an important feature added. Not entirely sure what NF next is, but you didn't make any new stats that didn't already exist.

Thanks for the review. ONVM_NF_ACTION_NEXT is used for a flow-table lookup action, as described [in the docs](https://github.com/sdnfv/openNetVM/blob/master/docs/NF_Dev.md)

File: onvm/onvm_nflib/onvm_nflib.c [hunk 1]
```diff
@@ -741,6 +741,10 @@ onvm_nflib_stop(struct onvm_nf_local_ctx *nf_local_ctx) {
 
         /* Terminate children */
         onvm_nflib_terminate_children(nf_local_ctx->nf);
+
+        /* Print statistics summary*/
```
Comment (koolzz) [inline onvm/onvm_nflib/onvm_nflib.c]: [nit] Space after summary

File: onvm/onvm_nflib/onvm_nflib.c [hunk 2]
```diff
@@ -1290,3 +1294,34 @@ init_shared_core_mode_info(uint16_t instance_id) {
...
+        const uint64_t act_drop = nfs[id].stats.act_drop;
+        const uint64_t act_next = nfs[id].stats.act_next;
+        const uint64_t act_buffer = nfs[id].stats.tx_buffer;
+        const uint64_t act_returned = nfs[id].stats.tx_returned;
+
+        /* Clear screen and move to top left */
+        printf("%s%s", clr, topLeft);
+        printf("NF activity summary\n");
+        printf("----------------------------------------------------\n");
+        printf("RX tot: %ld\n", rx);
```
Comment (koolzz) [inline onvm/onvm_nflib/onvm_nflib.c]: Maybe expand the tot-> total? As nf terminated we can use more screen space and it would look nicer.

**Comment by dennisafa** [comment]
Added more descriptive output as per your suggestion @koolzz should be good to review

**Comment by koolzz** [comment]
@onvm .

**Comment by onvm** [comment]
> @onvm .

## CI Message

Your results will arrive shortly

**Comment by onvm** [comment]
> @onvm .

## CI Message

Error: ERROR: Failed to post results to GitHub

**Comment by dennisafa** [comment]
> > @onvm .
> 
> ## CI Message
> Error: ERROR: Failed to post results to GitHub

🤔

**Comment by dennisafa** [comment]
> > @onvm .
> 
> ## CI Message
> Error: ERROR: Failed to post results to GitHub

🤔

**Comment by kevindweb** [comment]
Ya so idk why it happened. I just looked at the logs. `HTTPSConnectionPool(host='api.github.com', port=443): Read timed out.` It seems to have worked itself out though?

**Comment by dennisafa** [comment]
> > @onvm .
> 
> ## CI Message
> Error: ERROR: Failed to post results to GitHub

🤔

> Ya so idk why it happened. I just looked at the logs. `HTTPSConnectionPool(host='api.github.com', port=443): Read timed out.` It seems to have worked itself out though?

sounds like a github problem anyway

**Comment by kevindweb** [comment]
> > > @onvm .
> > 
> > 
> > ## CI Message
> > Error: ERROR: Failed to post results to GitHub
> 
> 🤔
> 
> > Ya so idk why it happened. I just looked at the logs. `HTTPSConnectionPool(host='api.github.com', port=443): Read timed out.` It seems to have worked itself out though?
> 
> sounds like a github problem anyway

You're right, somehow connecting to Github API took more than 10 seconds? But then somehow we actually approved the change in the end anyways?

**Comment by kevindweb** [comment]
> > > @onvm .
> > 
> > 
> > ## CI Message
> > Error: ERROR: Failed to post results to GitHub
> 
> 🤔
> 
> > Ya so idk why it happened. I just looked at the logs. `HTTPSConnectionPool(host='api.github.com', port=443): Read timed out.` It seems to have worked itself out though?
> 
> sounds like a github problem anyway

You're right, somehow connecting to Github API took more than 10 seconds? But then somehow we actually approved the change in the end anyways?

**Comment by kevindweb** [comment]
Ok ya Github is having problems. It just posted my comment twice in a row

**Comment by kevindweb** [comment]
Screen Shot 2020-01-19 at 4 11 51 PM

**Comment by twood02** [comment]
@dennisafa think about how this should print out if we are in the CSV format output mode -- maybe just print one line with these stats in CSV format with a header listing the fields

**Comment by koolzz** [comment]
@dennisafa, I agree with the above, implement and I'll merge.

**Comment by twood02** [comment]
@dennisafa is this still in progress or ready for review?

**Comment by dennisafa** [comment]
> @dennisafa is this still in progress or ready for review?

still tuning it, it will be done by this weekend.

**Comment by dennisafa** [comment]
in addition to printing to the terminal, a csv file is created in the nf's directory with its stats. i thought this might be better than setting a flag, since it would be an arg change for each nf

**Comment by dennisafa** [comment]
good to review @sdnfv/active_maintainers

**Comment by kevindweb** [comment]
@onvm test this stale review

**Comment by onvm** [comment]
> @onvm test this stale review

## CI Message

Your results will arrive shortly

File: onvm/onvm_nflib/onvm_nflib.c [hunk 3]
```diff
@@ -741,6 +741,9 @@ onvm_nflib_stop(struct onvm_nf_local_ctx *nf_local_ctx) {
 
         /* Terminate children */
         onvm_nflib_terminate_children(nf_local_ctx->nf);
+        /* Print statistics summary*/
```
Comment (kevindweb) [inline onvm/onvm_nflib/onvm_nflib.c]:
```suggestion

        /* Print statistics summary */
```

Small nit about comment `*/`. Also thought there should be a line between the call and comment? If not, that's ok too.

**Comment by kevindweb** [comment]
So much cleaner 👍

**Comment by kevindweb** [comment]
@onvm check this?

**Comment by onvm** [comment]
> @onvm check this?

## CI Message

Your results will arrive shortly

**Comment by bdevierno1** [comment]
Tested and got the correct desired output. Looks good. I decided to double check the output of the file with the linter on my terminal. 
*python --version
Python 2.7.16 
python style/gwclint.py --verbose=4 onvm/onvm_nflib/onvm_nflib.c
onvm/onvm_nflib/onvm_nflib.c:1349:  Almost always, snprintf is better than strcpy  [runtime/printf] [4]
onvm/onvm_nflib/onvm_nflib.c:1354:  Almost always, snprintf is better than strcat  [runtime/printf] [4]*
I understand from line 1343 you are getting the total length of the string plus the null character therefore, this will not make undefined behavior unlikely due the source size being greater than the destination. However, will using **strncat** may be a bit safer and will help remove these errors from the linter? If I am misunderstanding or using the linter incorrectly you can drop me a pm.

**Comment by dennisafa** [comment]
> Tested and got the correct desired output. Looks good. I decided to double check the output of the file with the linter on my terminal.
> _python --version Python 2.7.16 python style/gwclint.py --verbose=4 onvm/onvm_nflib/onvm_nflib.c onvm/onvm_nflib/onvm_nflib.c:1349: Almost always, snprintf is better than strcpy [runtime/printf] [4] onvm/onvm_nflib/onvm_nflib.c:1354: Almost always, snprintf is better than strcat [runtime/printf] [4]_
> I understand from line 1343 you are getting the total length of the string plus the null character therefore, this will not make undefined behavior unlikely due the source size being greater than the destination. However, will using **strncat** may be a bit safer and will help remove these errors from the linter?

Good catch, fixed

**Pull request merged**